File: data/sandy_mura.py
# -*- coding: utf-8 -*-
"""
Created on Fri Jul  1 12:16:06 2022

@author: F1_001
"""

import numpy as np
import openpyxl as xl
import pandas as pd
import statistics
import os,re,math,time

#保存路径设置
filesave2path = r"D:\data\LTPS_Tandem_15d5\puc\tet.xlsx"

#打开路径
filepath = r"D:\data\LTPS_Tandem_15d5\puc\ResultData"
data_list = os.listdir(filepath)


wb = xl.Workbook()
wb_sht = wb.active
wb_sht.title = 'sandy_mura'
wb_sht.cell(1, 2).value = 'min'
wb_sht.cell(1, 3).value = 'max'
wb_sht.cell(1, 4).value = 'avg'
wb_sht.cell(1, 5).value = 'std'
wb_sht.cell(1, 6).value = 'std/avg'

#2560*1600 max col=2460，max raw =1600
raw_start,raw_end = 600,1000
col_start,col_end = 1080,1480

wb_sht.cell(1, 8).value = str(str(raw_start)+'~'+str(raw_end)+','+str(col_start)+'~'+str(col_end))

time0 = time.time()
c=0
for each_data in data_list:
    each_data_path = os.path.join(filepath, each_data)
    #print(each_data)
    if 'PIEs_' in each_data:
        flg = "(.*).csv"
        name = re.compile(flg).findall(each_data)

        data = pd.read_csv(each_data_path,header=None)
        dt = []
        for i in range(raw_start,raw_end):
            for j in range(col_start,col_end):
                dt.append(data.iloc[[i],[j]].values[0][0])
        #min_Lv = min(dt)
        min_Lv = np.min(dt)
        max_Lv = np.max(dt)
        avg_Lv = np.mean(dt)
        #avg_Lv = sum(dt)/len(dt)
        std = statistics.pstdev(dt)
        std_avg = std/avg_Lv
        wb_sht.cell(c+2, 1).value = name[0]
        wb_sht.cell(c+2, 2).value = min_Lv
        wb_sht.cell(c+2, 3).value = max_Lv
        wb_sht.cell(c+2, 4).value = avg_Lv
        wb_sht.cell(c+2, 5).value = std
        wb_sht.cell(c+2, 6).value = std_avg
        
    c+=1
wb.save(filesave2path)
time1 = time.time()
t=time1-time0
print("total time = ",t)            

File: data/Sandy_mura.jl
import CSV,XLSX
using DataFrames,Statistics,Dates
# DelimitedFiles,StringEncodings

savepath = raw"D:\data\FPGA_DEMURA\Vth\vthon_ttt2.xlsx"
filepath = raw"D:\data\FPGA_DEMURA\Vth\vthon"

datalist = readdir(filepath)


# raw_start,raw_end = 750,850
# col_start,col_end = 1230,1330
function Sandy(filename)
    file = joinpath(filepath,filename)
    data = CSV.read(file,DataFrame,header=0)
    size_ = size(data)
    raw_start = Int64(size_[1]/2-50)
    raw_end = Int64(size_[1]/2+50)
    col_start = Int64(size_[2]/2-50)
    col_end = Int64(size_[2]/2+50)
    # println(raw_start,col_start)
    dt = []
    for i in raw_start+2:raw_end-2
        for j in col_start+2:col_end-2
            block = []
            a = data[i,j]
            for m in i-2:i+2
                for n in j-2:j+2
                    if m != i & n != j
                        b = data[m,n]
                        d = 1-min(a,b)/max(a,b)
                        # push!(dt,d)
                        push!(block,d)
                    end
                end
            end
            dmax = findmax(block)[1]
            push!(dt,dmax)
        end
    end
    return dt
end


t1 = Dates.now()
XLSX.openxlsx(savepath, mode="w") do xf
    sheet = xf[1]
    XLSX.rename!(sheet, "sandy")
    sheet[1,2] = "min"
    sheet[1,3] = "max"
    sheet[1,4] = "avg"
    sheet[1,5] = "std"
    sheet[1,6] = "std/avg"
    # sheet["A1"] = "this is a"
    # sheet["A2"] = "new file"
    # sheet["A3"] = "1111"
    # sheet["A4"] = 10000
    # sheet["A5"] = collect(1:5) 
    # sheet["B1", dim=1] = collect(1:4)
    # sheet["A7:C9"] = [ 1 2 3 ; 4 5 6 ; 7 8 9 ]


raw_start,raw_end = 750,850
col_start,col_end = 1230,1330

c = 2
for filename in datalist
    if occursin("csv",filename)
        dt = Sandy(filename)
        min = findmin(dt)
        max = findmax(dt)
        avg = mean(dt)
        std = Statistics.std(dt)
        std_avg = std/avg
        name = match(r"(.*).csv",filename)
        sheet[c,1] = string(name[1])
        sheet[c,2] = min[1]
        sheet[c,3] = max[1]
        sheet[c,4] = avg
        sheet[c,5] = std
        sheet[c,6] = std/avg
        c += 1
    end
end
end
t2 = Dates.now()
println(t2-t1)
